ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/Authentication.kt

```diff
@@ -5,262 +5,80 @@
-
-        for (challenge in challenges) {
-            challengePipeline.intercept(ChallengePhase) {
-                challenge(it)
-                if (it.completed) {
-                    finish() // finish challenge pipeline if it has been completed
-                }
-            }
-        }
-
-        for (challenge in context.challenge.errorChallenges) {
-            challengePipeline.intercept(ChallengePhase) {
-                challenge(it)
-                if (it.completed) {
-                    finish() // finish challenge pipeline if it has been completed
-                }
-            }
-        }
-
-        for (error in context.allErrors) {
-            challengePipeline.intercept(ChallengePhase) {
-                if (!it.completed) {
-                    logger.trace("Responding unauthorized because of error ${error.message}")
-                    call.respond(UnauthorizedResponse())
-                    it.complete()
-                    finish()
-                }
-            }
-        }
-
-        val challenge = challengePipeline.execute(call, context.challenge)
-        if (challenge.completed) {
-            finish() // finish authentication pipeline if challenge has been completed
-        }
-    }
-
-    private suspend fun PipelineContext<Unit, ApplicationCall>.processAuthentication(
-        call: ApplicationCall,
-        authenticationContext: AuthenticationContext,
-        configurations: List<AuthenticationProvider>,
-        defaultPipeline: AuthenticationPipeline
-    ) {
-        val firstSkippedIndex = configurations.indexOfFirst {
-            it.skipWhen.any { skipCondition -> skipCondition(call) }
-        }
-
-        if (firstSkippedIndex == -1) { // all configurations are unskipped - apply the whole prebuilt pipeline
-            defaultPipeline.execute(call, authenticationContext)
-        } else if (firstSkippedIndex == 0 && configurations.size == 1) {
-            return // the only configuration is skipped - fast-path return
-        } else {
-            // rebuild pipeline to skip particular auth methods
-
-            val child = AuthenticationPipeline(defaultPipeline.developmentMode)
-            var applicableCount = 0
-
-            for (idx in 0 until firstSkippedIndex) {
-                child.merge(configurations[idx].pipeline)
-                applicableCount++
-            }
-
-            for (idx in firstSkippedIndex + 1 until configurations.size) {
-                val provider = configurations[idx]
-                if (provider.skipWhen.none { skipCondition -> skipCondition(call) }) {
-                    child.merge(provider.pipeline)
-                    applicableCount++
-                }
-            }
-
-            if (applicableCount > 0) {
-                child.execute(call, authenticationContext)
-            }
-        }
+    public companion object : ApplicationPlugin<Application, AuthenticationConfig, Authentication> {
```

Comment:
Could you tell me why do we still need this in the new API?

Reply:
This plugin is application level and keeps all installed providers. It needs to be in the old API, because we need to access its instance.